crates/bevy_input_focus/src/tab_navigation.rs

```diff
@@ -0,0 +1,336 @@
+///
+/// Note that you must also add the [`TabGroup`] component to the entity's ancestor in order
+/// for this component to have any effect.
+#[derive(Debug, Default, Component, Copy, Clone)]
+pub struct TabIndex(pub i32);
+
+/// Indicates that this widget should automatically receive focus when it's added.
+#[derive(Debug, Default, Component, Copy, Clone)]
+pub struct AutoFocus;
+
+/// A component used to mark a tree of entities as containing tabbable elements.
+#[derive(Debug, Default, Component, Copy, Clone)]
+pub struct TabGroup {
+    /// The order of the tab group relative to other tab groups.
+    pub order: i32,
+
+    /// Whether this is a 'modal' group. If true, then tabbing within the group (that is,
+    /// if the current focus entity is a child of this group) will cycle through the children
+    /// of this group. If false, then tabbing within the group will cycle through all non-modal
+    /// tab groups.
+    pub modal: bool,
+}
+
+impl TabGroup {
+    /// Create a new tab group with the given order.
+    pub fn new(order: i32) -> Self {
+        Self {
+            order,
+            modal: false,
+        }
+    }
+
+    /// Create a modal tab group.
+    pub fn modal() -> Self {
+        Self {
+            order: 0,
+            modal: true,
+        }
+    }
+}
+
+/// Navigation action for tabbing.
+pub enum NavAction {
+    /// Navigate to the next focusable entity, wrapping around to the beginning if at the end.
+    Next,
+    /// Navigate to the previous focusable entity, wrapping around to the end if at the beginning.
+    Previous,
+    /// Navigate to the first focusable entity.
+    First,
+    /// Navigate to the last focusable entity.
+    Last,
+}
+
+/// An injectable helper object that provides tab navigation functionality.
+#[doc(hidden)]
+#[derive(SystemParam)]
+#[allow(clippy::type_complexity)]
+pub struct TabNavigation<'w, 's> {
+    // Query for tab groups.
+    tabgroup_query: Query<'w, 's, (Entity, &'static TabGroup, &'static Children)>,
+    // Query for tab indices.
+    tabindex_query: Query<
+        'w,
+        's,
+        (Entity, Option<&'static TabIndex>, Option<&'static Children>),
+        Without<TabGroup>,
+    >,
+    // Query for parents.
+    parent_query: Query<'w, 's, &'static Parent>,
+}
+
+impl TabNavigation<'_, '_> {
+    /// Navigate to the next focusable entity.
+    ///
+    /// Arguments:
+    /// * `focus`: The current focus entity, or `None` if no entity has focus.
+    /// * `action`: Whether to select the next, previous, first, or last focusable entity.
+    ///
+    /// If no focusable entities are found, then this function will return either the first
+    /// or last focusable entity, depending on the direction of navigation. For example, if
+    /// `action` is `Next` and no focusable entities are found, then this function will return
+    /// the first focusable entity.
+    pub fn navigate(&self, focus: Option<Entity>, action: NavAction) -> Option<Entity> {
```

Comment:
I'm not sure if it would fit with the goal of this param, but it could read and write directly into `ResMut<InputFocus>` so the `focus` param is not necessary.

Reply:
I thought about that, but doing it this way makes it a bit more flexible, and also easier to test.